File: 145. Binary Tree Postorder Traversal.cpp
#include <iostream>
#include <sstream>
#include <algorithm>
#include <queue>
#include <stack>
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* node) {
        // https://stackoverflow.com/questions/36384599/can-we-use-morris-traversal-for-postorder
        TreeNode virtualRoot(0);
        virtualRoot.left = node;
        node = &virtualRoot;
        vector<int> res;
        TreeNode *pre, *first, *mid, *last;
        while (node) {
            if (!node->left) {
                node = node->right;
            }
            else {  // with left child
                pre = node->left;
                while (pre->right && pre->right != node) {
                    pre = pre->right;
                }
                if (!pre->right) {
                    pre->right = node;
                    node = node->left;
                }
                else {
                    first = node;
                    mid = node->left;
                    while (mid != node) {
                        last = mid->right;
                        mid->right = first;
                        first = mid;
                        mid = last;
                    }
                    first = node;
                    mid = pre;
                    while (mid != node) {
                        res.push_back(mid->val);
                        last = mid->right;
                        mid->right = first;
                        first = mid;
                        mid = last;
                    }
                    pre->right = NULL;  // make right pointer null again
                    node = node->right;
                }
            }
        }
        return res;
    }
    
    vector<int> postorderTraversalWithStack(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> s;
        TreeNode *topNode, *lastNodeVisited;
        while (!s.empty() || root) {
            if (root) {
                s.push(root);
                root = root->left;
            }
            else {
                topNode = s.top();
                if (topNode->right && lastNodeVisited != topNode->right) {
                    root = topNode->right;
                }
                else {
                    res.push_back(topNode->val);
                    lastNodeVisited = topNode;
                    s.pop();
                    root = NULL;
                }
            }
        }
        return res;
    }
};

void trimLeftTrailingSpaces(string &input) {
    input.erase(input.begin(), find_if(input.begin(), input.end(), [](int ch) {
        return !isspace(ch);
    }));
}

void trimRightTrailingSpaces(string &input) {
    input.erase(find_if(input.rbegin(), input.rend(), [](int ch) {
        return !isspace(ch);
    }).base(), input.end());
}

TreeNode* stringToTreeNode(string input) {
    trimLeftTrailingSpaces(input);
    trimRightTrailingSpaces(input);
    input = input.substr(1, input.length() - 2);
    if (!input.size()) {
        return nullptr;
    }

    string item;
    stringstream ss;
    ss.str(input);

    getline(ss, item, ',');
    TreeNode* root = new TreeNode(stoi(item));
    queue<TreeNode*> nodeQueue;
    nodeQueue.push(root);

    while (true) {
        TreeNode* node = nodeQueue.front();
        nodeQueue.pop();

        if (!getline(ss, item, ',')) {
            break;
        }

        trimLeftTrailingSpaces(item);
        if (item != "null") {
            int leftNumber = stoi(item);
            node->left = new TreeNode(leftNumber);
            nodeQueue.push(node->left);
        }

        if (!getline(ss, item, ',')) {
            break;
        }

        trimLeftTrailingSpaces(item);
        if (item != "null") {
            int rightNumber = stoi(item);
            node->right = new TreeNode(rightNumber);
            nodeQueue.push(node->right);
        }
    }
    return root;
}

void println(const vector<int>& v) {
    for (int i: v) cout << i << ' ';
    cout << endl;
}

int main() {
    Solution tra;
    string nodelist = "[3,9,20,11,23,15,7]";
    TreeNode* root = stringToTreeNode(nodelist);
    vector<int> res;
    res = tra.postorderTraversal(root);
    println(res);
    return 0;
}

File: _145_Binary_Tree_Postorder_Traversal.java
import java.util.*;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        Stack<TreeNode> s = new Stack<>();
        TreeNode topNode = null, lastNodeVisited = null;
        while (!s.empty() || root != null) {
            if (root != null) {
                s.push(root);
                root = root.left;
            }
            else {
                topNode = s.peek();
                if (topNode.right != null && topNode.right != lastNodeVisited) {
                    root = topNode.right;
                }
                else {
                    ans.add(topNode.val);
                    lastNodeVisited = topNode;
                    s.pop();
                }
            }
        }
        return ans;
    }
}

public class _145_Binary_Tree_Postorder_Traversal {
    public static void main(String[] args) {
        String line = "[1,2,3,4,5,6,7,8,9,10]";
        TreeNode root = stringToTreeNode(line);
            
        List<Integer> ret = new Solution().postorderTraversal(root);
        
        String out = integerArrayListToString(ret);
        
        System.out.println(out);
    }

    public static TreeNode stringToTreeNode(String input) {
        input = input.trim();
        input = input.substring(1, input.length() - 1);
        if (input.length() == 0) {
            return null;
        }
    
        String[] parts = input.split(",");
        String item = parts[0];
        TreeNode root = new TreeNode(Integer.parseInt(item));
        Queue<TreeNode> nodeQueue = new LinkedList<>();
        nodeQueue.add(root);
    
        int index = 1;
        while(!nodeQueue.isEmpty()) {
            TreeNode node = nodeQueue.remove();
    
            if (index == parts.length) {
                break;
            }
    
            item = parts[index++];
            item = item.trim();
            if (!item.equals("null")) {
                int leftNumber = Integer.parseInt(item);
                node.left = new TreeNode(leftNumber);
                nodeQueue.add(node.left);
            }
    
            if (index == parts.length) {
                break;
            }
    
            item = parts[index++];
            item = item.trim();
            if (!item.equals("null")) {
                int rightNumber = Integer.parseInt(item);
                node.right = new TreeNode(rightNumber);
                nodeQueue.add(node.right);
            }
        }
        return root;
    }
    
    public static String integerArrayListToString(List<Integer> nums, int length) {
        if (length == 0) {
            return "[]";
        }
    
        String result = "";
        for(int index = 0; index < length; index++) {
            Integer number = nums.get(index);
            result += Integer.toString(number) + ", ";
        }
        return "[" + result.substring(0, result.length() - 2) + "]";
    }
    
    public static String integerArrayListToString(List<Integer> nums) {
        return integerArrayListToString(nums, nums.size());
    }
}
